compiler/src/statement/assign/assignee/array_range_index.rs

```diff
@@ -0,0 +1,107 @@
+
+impl<'a, F: PrimeField, G: GroupType<F>> ConstrainedProgram<'a, F, G> {
+    pub(super) fn resolve_target_access_array_range<'b, CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        mut context: ResolverContext<'a, 'b, F, G>,
+        start: Option<&'a Expression<'a>>,
+        stop: Option<&'a Expression<'a>>,
+    ) -> Result<(), StatementError> {
+        let start_index = start
+            .map(|start| self.enforce_index(cs, start, &context.span))
+            .transpose()?
+            .map(|x| {
+                x.to_usize()
+                    .ok_or_else(|| StatementError::array_assign_index_const(&context.span))
+            })
+            .transpose()?;
+        let stop_index = stop
+            .map(|stop| self.enforce_index(cs, stop, &context.span))
+            .transpose()?
+            .map(|x| {
+                x.to_usize()
+                    .ok_or_else(|| StatementError::array_assign_index_const(&context.span))
+            })
+            .transpose()?;
+        let start_index = start_index.unwrap_or(0);
+
+        if context.input.len() == 1 {
+            // not a range of a range
+            match context.input.remove(0) {
+                ConstrainedValue::Array(old) => {
+                    let stop_index = stop_index.unwrap_or(old.len());
+                    Self::check_range_index(start_index, stop_index, old.len(), &context.span)?;
+
+                    if context.remaining_accesses.is_empty() {
+                        let target_values = match context.target_value {
+                            ConstrainedValue::Array(x) => x,
+                            _ => unimplemented!(),
+                        };
+
+                        for (target, target_value) in old[start_index..stop_index].iter_mut().zip(target_values) {
+                            context.target_value = target_value;
+                            self.enforce_assign_context(cs, &context, target)?;
+                        }
+                    } else {
+                        context.input = old[start_index..stop_index].iter_mut().collect();
+                        self.resolve_target_access(cs, context)?;
+                    }
+                    Ok(())
+                }
+                _ => Err(StatementError::array_assign_index(&context.span)),
+            }
+        } else {
+            // range of a range
+            let stop_index = stop_index.unwrap_or(context.input.len());
+            Self::check_range_index(start_index, stop_index, context.input.len(), &context.span)?;
+
+            context.input = context
+                .input
+                .into_iter()
+                .skip(start_index)
+                .take(stop_index - start_index)
+                .collect();
+            if context.remaining_accesses.is_empty() {
+                let target_values = match context.target_value {
+                    ConstrainedValue::Array(x) => x,
+                    _ => unimplemented!(),
```

Comment:
Can this case ever be reached?

Reply:
nope, would be a critical implementation failure